---
title: "Multiome Process3"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document
---

This R script will do the peak recalling, batch correction and process cluster markers, it requires 4 inputs: 'after_QC.RDS', directory of macs2, ATAC_seq, and 'annotations.RDS'. After running, 'after_link_peaks.RDS' and 'cluster_markers.RDS' will be saved for cell type annotation and following analysis. 

```{r,echo=F,message=F,warning=F}
library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(dplyr)
library(ggplot2)
library(SoupX)
library(Matrix)
library(scDblFinder)
library(SeuratData)
library(SeuratDisk)
library(harmony)
library(BSgenome.Hsapiens.UCSC.hg38)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(stringr)
library(gridExtra)
library(grid)
library(yaml)
library(config)

config = yaml.load_file("config.yml")
Sys.setenv(R_CONFIG_ACTIVE = "default")
dir_macs2= config::get("dir_macs2")
dir_annotations= config::get("dir_annotations")
dir_ATAC = config::get("dir_ATAC")
output_dir = config::get("output_dir")
dir_after_QC = paste(output_dir,'/','after_QC.RDS', sep="")
```


## Peak Recalling

The set of peaks identified using Cellranger often merges distinct peaks that are close together. This can create a problem for certain analyses, particularly motif enrichment analysis and peak-to-gene linkage. To identify a more accurate set of peaks, we can call peaks using MACS2. Here we call peaks on all cells together, but we could identify peaks for each group of cells separately by setting the 'group.by' parameter, and this can help identify peaks specific to rare cell populations.


```{r,echo=F,message=F,warning=F}
gene.multiome.seurat.filtered <- readRDS(dir_after_QC)
annotations <- readRDS(dir_annotations)
frag.file <- dir_ATAC

DefaultAssay(gene.multiome.seurat.filtered) <- "ATAC"
temp.peaks <- CallPeaks(gene.multiome.seurat.filtered, macs2.path = dir_macs2)

# remove peaks on nonstandard chromosomes and in genomic blacklist regions
temp.peaks <- keepStandardChromosomes(temp.peaks, pruning.mode = "coarse")
temp.peaks <- subsetByOverlaps(x = temp.peaks, ranges = blacklist_hg38_unified, invert = TRUE)

# quantify counts in each peak

temp.macs2.counts <- FeatureMatrix(fragments = Fragments(gene.multiome.seurat.filtered),  features = temp.peaks,  cells = colnames(gene.multiome.seurat.filtered))

# create a new assay using the MACS2 peak set and add it to the Seurat object

gene.multiome.seurat.filtered[["peaks"]] <- CreateChromatinAssay( counts = temp.macs2.counts,  fragments = frag.file,  annotation = annotations)
```


## Gene expression and DNA accessibility data processing 

We next perform normalization and pre-processing and dimensional reduction on both assays independently, using standard approaches for RNA and ATAC-seq data.

```{r,echo=F,message=F,warning=F}
# RNA analysis
DefaultAssay(gene.multiome.seurat.filtered) <- "RNA"
gene.multiome.seurat.filtered <- SCTransform(gene.multiome.seurat.filtered, verbose = FALSE) %>% RunPCA() 

# ATAC analysis
# We exclude the first dimension as this is typically correlated with sequencing depth
DefaultAssay(gene.multiome.seurat.filtered) <- "peaks"
gene.multiome.seurat.filtered <- RunTFIDF(gene.multiome.seurat.filtered)
gene.multiome.seurat.filtered <- FindTopFeatures(gene.multiome.seurat.filtered, min.cutoff = 'q0')
gene.multiome.seurat.filtered <- RunSVD(gene.multiome.seurat.filtered)
```


## Batch Correction

Batch correction was performed here.

```{r,echo=F,warning=F}

Idents(gene.multiome.seurat.filtered) = gene.multiome.seurat.filtered$batch 

DefaultAssay(gene.multiome.seurat.filtered) <- "RNA"

gene.multiome.seurat.filtered =RunHarmony(gene.multiome.seurat.filtered, group.by.vars=c("batch"),reduction="pca", reduction.save= "harmony.pca",assay.use="SCT")

DefaultAssay(gene.multiome.seurat.filtered) <- "peaks"

gene.multiome.seurat.filtered =RunHarmony(gene.multiome.seurat.filtered, group.by.vars=c("batch"),reduction="lsi", reduction.save= "harmony.lsi",assay.use="peaks", project.dim=F)

```


```{r,echo=F,message=F,warning=F}
### integrated

### We calculate a WNN graph, representing a weighted combination of RNA and ATAC-seq modalities. We use this graph for UMAP visualization and clustering

gene.multiome.seurat.filtered <- FindMultiModalNeighbors(gene.multiome.seurat.filtered, reduction.list = list("harmony.pca", "harmony.lsi"), dims.list = list(1:50, 2:50))

gene.multiome.seurat.filtered <- RunUMAP(gene.multiome.seurat.filtered, dims = 1:50, reduction.name = 'umap.harmony.pca',  reduction="harmony.pca")

gene.multiome.seurat.filtered <- RunUMAP(gene.multiome.seurat.filtered, dims = 1:50, reduction.name = 'umap.harmony.lsi',  reduction="harmony.lsi")

gene.multiome.seurat.filtered <- RunUMAP(gene.multiome.seurat.filtered, nn.name = "weighted.nn", reduction.name = "umap.harmony.wnn")

gene.multiome.seurat.filtered <- FindClusters(gene.multiome.seurat.filtered, graph.name = "wsnn", algorithm = 3, verbose = FALSE)

saveRDS(gene.multiome.seurat.filtered,file = paste(output_dir,'/','after_batch_correction.RDS', sep=""))
```



```{r,echo=F,message=F,warning=F}
### marker genes 
DefaultAssay(gene.multiome.seurat.filtered) <- "SCT"

# This loop just runs the FindMarkers function on all of the clusters

SSc.gene.clustermarkers <- list()
Idents(gene.multiome.seurat.filtered) <- gene.multiome.seurat.filtered$seurat_clusters

for (i in levels(gene.multiome.seurat.filtered[["seurat_clusters"]][[1]])) {
  
  if ((length(which(gene.multiome.seurat.filtered$seurat_clusters == i)) > 3) &
  (length(which(gene.multiome.seurat.filtered$seurat_clusters != i)) > 3)) {
    
    SSc.gene.clustermarkers[[i]] <- FindMarkers(gene.multiome.seurat.filtered,ident.1 = i,min.pct = 0.1, test.use = "wilcox" , logfc.threshold=0.25, only.pos=T)
  
  }else{
    SSc.gene.clustermarkers[[i]] <- c()
  }
  
}

# lapply(
#   levels(gene.multiome.seurat.filtered[["seurat_clusters"]][[1]]),function(x)FindMarkers(gene.multiome.seurat.filtered,ident.1 = x,min.pct = 0.1, test.use = "wilcox" , logfc.threshold=0.25, only.pos=T)) -> SSc.gene.clustermarkers

saveRDS(SSc.gene.clustermarkers,file = paste(output_dir,'/','cluster_markers.RDS', sep=""))
```


```{r,echo=F,message=F,warning=F}
DefaultAssay(gene.multiome.seurat.filtered) <- "peaks"

# first compute the GC content for each peak
gene.multiome.seurat.filtered <- RegionStats(gene.multiome.seurat.filtered, genome = BSgenome.Hsapiens.UCSC.hg38)

# link peaks to genes
gene.multiome.seurat.filtered <- LinkPeaks(  object = gene.multiome.seurat.filtered,  peak.assay = "peaks",  expression.assay = "SCT")

saveRDS(gene.multiome.seurat.filtered,file = paste(output_dir,'/','after_link_peaks.RDS', sep=""))
```

#UMAP

```{r,echo=F,message=F,warning=F,fig.height=5, fig.width=13}

grid_arrange_shared_legend <- function(..., nrow = 1, ncol = length(list(...)), position = c("bottom", "right")) {

  plots <- list(...)
  position <- match.arg(position)
  g <- ggplotGrob(plots[[1]] + theme(legend.position = position))$grobs
  legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
  lheight <- sum(legend$height)
  lwidth <- sum(legend$width)
  gl <- lapply(plots, function(x) x + theme(legend.position = "none"))
  gl <- c(gl, nrow = nrow, ncol = ncol)

  combined <- switch(position,
                     "bottom" = arrangeGrob(do.call(arrangeGrob, gl),
                                            legend,
                                            ncol = 1,
                                            heights = unit.c(unit(1, "npc") - lheight, lheight)),
                     "right" = arrangeGrob(do.call(arrangeGrob, gl),
                                           legend,
                                           ncol = 2,
                                           widths = unit.c(unit(1, "npc") - lwidth, lwidth)))
  grid.newpage()
  grid.draw(combined)

}

p1 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.pca", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle(paste("RNA (harmony)"))
p2 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.lsi", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle(paste("ATAC (harmony)"))
p3 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.wnn", label = TRUE, label.size = 2.5, repel = TRUE) + ggtitle(paste("WNN (harmony)"))

grid_arrange_shared_legend(p1, p2, p3, nrow = 1, ncol = 3,position="right")

p1 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.pca", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'batch') + ggtitle(paste("RNA (harmony) by batch"))
p2 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.lsi", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'batch') + ggtitle(paste("ATAC (harmony) by batch"))
p3 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.wnn", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'batch') + ggtitle(paste("WNN (harmony) by batch"))

grid_arrange_shared_legend(p1, p2, p3, nrow = 1, ncol = 3,position="right")
```













