---
title: "Multiome Preprocessing"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document

---

This R script will do the pre-processing covering ambient RNA removal, adding ATAC-seq data, and doublet removal. There are five required inputs: RNA-seq data and ATAC-seq data, forceOption, sample info file, and clustering info from cellranger. After running, two RDS 'before_QC.RDS' and 'annotations' will be saved for following analysis.

# Check whether the order of sample input matches the aggr input file

```{r,echo=F,warning=F}
library(yaml)
library(config)
config = yaml.load_file("config.yml")
Sys.setenv(R_CONFIG_ACTIVE = "default")
dir_gene = config::get("dir_gene")
dir_ATAC = config::get("dir_ATAC")
forceOption = config::get("forceOption")
batch_data = config::get("batch_data")
dir_cluster = config::get("dir_cluster")
aggr_input = config::get("aggr_input")
output_dir = config::get("output_dir")
aggr <- read.csv(aggr_input)
temp <- read.csv(batch_data)
if (aggr[, 1] != temp[, 1]) {
  print("The order of the sample input file doesn't match the order of samples from aggr input file!")
  knitr::knit_exit()
}
```

```{r,echo=F,message=F,warning=F}
library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(dplyr)
library(ggplot2)
library(SoupX)
library(Matrix)
library(scDblFinder)
library(harmony)
library(BSgenome.Hsapiens.UCSC.hg38)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(stringr)
library(knitr)
library(data.table)
library(kableExtra)
```

# Get the batch and sample information

```{r,echo=F,message=F,warning=F}
#Seurat has an easy function to load in data generated by CellRanger. This funciton will read those files into a single sparse matrix variable in R with genes as rownames and cell barcodes at colnames.
gene.multiome <- Read10X_h5(dir_gene)
```


```{r,echo=F,message=F,warning=F}
## creates a Seurat object based on the RNA-seq data (unique)
gene.sample <- unlist(lapply(strsplit(colnames(gene.multiome$`Gene Expression`),split="-"),function(x){x[2]}))

## temp.gene.Seurat contains an assay storing RNA measurement
temp.gene.seurat <- list()
for (i in unique(gene.sample)){
  temp.gene.seurat[[i]] <- CreateSeuratObject(counts=gene.multiome$`Gene Expression`[,gene.sample==i])
}

batch_info <- strsplit(colnames(gene.multiome$`Gene Expression`), split = "-")
batch_info <- as.data.frame(matrix(unlist(batch_info), ncol = 2, byrow = T))
regexp <- "[[:digit:]]+"
temp$batch <- str_extract(temp$Batch, regexp)
Sample <- seq(1:nrow(temp))
temp2 <- cbind(Sample, temp)
temp2 <- setDT(temp2)
kable(temp2) %>% kableExtra::kable_styling(latex_options = c("scale_down", "HOLD_position"))
```

Here, we print out the annotations to check whether the seqnames begins with "chr" or not

```{r,echo=F,message=F,warning=F}
## Pulls the transcript information for all chromosomes from an EnsDb object. 
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
## Change the style of the seqlevels in which results returned from the EnsDb object are encoded ('UCSC')
seqlevelsStyle(annotations) <- 'UCSC'

annotations
```

## Remove ambient RNA

```{r,echo=F,warning=F}

#load in the cluster info from cellranger
cluster_info <- read.csv(dir_cluster)

sample_info <- strsplit(as.character(cluster_info$Barcode), "-")

sample_info <- as.data.frame(matrix(unlist(sample_info), ncol = 2, byrow = T))

cluster_info$sample <- sample_info$V2

tempout <- list()

for (i in (1:length(temp.gene.seurat))){
  
  i=as.character(i)
  # build the count matrix
  tempc <- GetAssayData(temp.gene.seurat[[i]], slot= "counts")
  
  # Creates a SoupChannel object
  tempSC <- SoupChannel(tempc, tempc, calcSoupProfile=F)
  
  tempSCprofile = data.frame(row.names = rownames(tempc), est = rowSums(tempc)/sum(tempc), counts = rowSums(tempc))
  
  tempSC = setSoupProfile(tempSC, tempSCprofile)
  
  # use cellranger clustering info
  
  tempSC <- setClusters(tempSC, setNames(subset(cluster_info,cluster_info$sample==i)$Cluster, subset(cluster_info,cluster_info$sample==i)$Barcode))
  
  cat(sprintf("\nFor sample %s:",i))
  
  # Estimating the contamination fraction, calculating ambient RNA profile
  tempSC <- autoEstCont(tempSC,doPlot = FALSE,forceAccept = forceOption)
  
  ### no need for integer
  ## After the level of background contamination has been estimated or specified for a channel, calculate the 
  ## resulting corrected count matrix with background contamination removed.
  tempout[[i]] <- adjustCounts(tempSC,roundToInt=F)
  
  cat(sprintf(paste("Proportion of ambient RNA removed (nCount):",1-sum(tempout[[i]])/sum(tempc),"\n",sep="")))
}
```


```{r,echo=F,message=F,warning=F}
## Sanity check
rowname.check = all(rownames(gene.multiome$`Gene Expression`) == rownames(tempout[[1]]))
if (rowname.check == FALSE) {
  print("Warning: the rownames are not consistant")
  knitr::knit_exit()
}
```

```{r,echo=F,message=F,warning=F}
## update the count matrix after RNA ambient removement
for (i in (1:length(tempout))){
  #print(i)
  gene.multiome$`Gene Expression`[,colnames(tempout[[i]])] <- tempout[[i]]
}
```

## Pre-doublet identification

Summary before pre-doublet identification:


```{r,echo=F,message=F,warning=F}
## before removing pre-doublet identification, how many ?? we have

# for (i in (1:length(temp.gene.seurat))){
#   print(c(i,dim(temp.gene.seurat[[i]])))
# }

tab2 <- seq(1:nrow(temp))
n.cells <- c()
for (i in (1:length(temp.gene.seurat))) {
  n.cells <- c(n.cells, dim(tempout[[i]])[2])
}
tab2 <- as.data.frame(tab2)
tab2$n.cells <- n.cells
colnames(tab2) <- c("sample", "n.cells")
tab2 <- setDT(tab2)
kable(tab2)

temp.singles <- c()

for (i in (1:length(temp.gene.seurat))){
  
  # print(i)
  # transform class for the purpose of using scDblFinder
  temp.sce <- as.SingleCellExperiment(temp.gene.seurat[[i]])
  
  #scDblFinder will add a number of columns to the colData of sce prefixed with ‘scDblFinder’
  temp.sce <- scDblFinder(temp.sce)
  temp.sce <- as.Seurat(temp.sce)
  a <- as.matrix(temp.sce$scDblFinder.class)
  temp.singles <- rbind(temp.singles,a)
  
}
```


The following table shows the summary of doublet and singlet.

```{r,echo=F,message=F,warning=F}
table(temp.singles)

## again, recreate Seurat object after ambient RNA removal and doublet identification
gene.multiome.seurat <- CreateSeuratObject(counts = gene.multiome$`Gene Expression`)
gene.multiome.seurat[["percent.mt"]] <- PercentageFeatureSet(gene.multiome.seurat, pattern = "^MT-")
```

## Add in the ATAC-seq data

In this step, we add in ATAC-seq data, and please notice that we only use peaks in standard chromosomes.

```{r,echo=F,message=F,warning=F}
## Add in the ATAC-seq data (only use peaks in standard chromosomes)
grange.counts <- StringToGRanges(rownames(gene.multiome$`Peaks`), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac_counts <- gene.multiome$`Peaks`[as.vector(grange.use), ]

frag.file <- dir_ATAC

gene.multiome.seurat[["ATAC"]] <- CreateChromatinAssay(
  counts = atac_counts,
  sep = c(":", "-"),
  genome = 'hg38',
  fragments = frag.file,
  min.cells = 10,
  annotation = annotations
)

Annotation(gene.multiome.seurat[["ATAC"]]) <- annotations

```

## Doublet removal

```{r,echo=F,message=F,warning=F}
## doublet removal based on above results
gene.multiome.seurat$singlet <- temp.singles[names(gene.multiome.seurat$orig.ident),1]

#add sample info
sample_info <- strsplit(colnames(gene.multiome.seurat),split="-")
sample_info <- as.data.frame(matrix(unlist(sample_info),ncol=2,byrow=T))

sample_info <- as.numeric(sample_info$V2)

gene.multiome.seurat$sample <- sample_info

for (i in 1:ncol(temp)) {
  
  cur_value <- as.numeric(sample_info)
  
  for (j in 1:nrow(temp)) {
    
    eval(parse(text=paste("cur_value <- replace(x <- cur_value, x == j,temp$",colnames(temp)[i], "[[j]])", sep = "")))

  }
 eval(parse(text=paste("gene.multiome.seurat$",colnames(temp)[i]," <- cur_value",sep="")))
   
}


Idents(gene.multiome.seurat) <- gene.multiome.seurat$sample

count_singlet <- c()
count_doublet <- c()
doublet_ratio <- c()
total_cells <- c()

for (i in (1:length(temp.gene.seurat))){

  tmp <- subset(gene.multiome.seurat, idents = i)
  tmp_singlet <- length(which(tmp$singlet == "singlet"))
  count_singlet <- c(count_singlet, tmp_singlet)
  tmp_doublet <- length(which(tmp$singlet == "doublet"))
  count_doublet <- c(count_doublet, tmp_doublet)
  doublet_ratio <- c(doublet_ratio, tmp_doublet / (tmp_doublet + tmp_singlet))
  total_cells <- c(total_cells, tmp_singlet + tmp_doublet)
}

tab3 <- cbind(seq(1:nrow(temp)), total_cells)
tab3 <- as.data.frame(tab3)
colnames(tab3) <- c("sample", "total_cells")
tab3$count_singlet <- count_singlet
tab3$count_doublet <- count_doublet
tab3$doublet_ratio <- doublet_ratio
tab3 <- setDT(tab3)
kable(tab3)



#filter out doublets
Idents(gene.multiome.seurat) <- gene.multiome.seurat$singlet
gene.multiome.seurat <- subset(gene.multiome.seurat,idents=c("singlet"))

print(paste("After doublet detection, the number of cells:", as.numeric(ncol(gene.multiome.seurat)),"; unique reads:",as.integer(sum(gene.multiome.seurat$nCount_RNA))))


print(paste("After doublet detection, the number of cells:", as.numeric(ncol(gene.multiome.seurat)),"; unique ATAC reads:",as.integer(sum(gene.multiome.seurat$nCount_ATAC))))
```


## Check MT percent after removement of ambient RNA and doublets

The following table provide some basic summary statistics for MT percent after removement of ambient RNA and doublets.

```{r,echo=F,message=F,warning=F}
summary(gene.multiome.seurat[["percent.mt"]][,1])

saveRDS(gene.multiome.seurat,file=paste(output_dir,'/','before_QC.RDS', sep=""))
saveRDS(annotations,file=paste(output_dir,'/','annotations.RDS', sep=""))
```














