---
title: "Multiome Analysis with Celltype"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: pdf_document
params:
  dir_config: F
---

This R script will do the UMAP, cell type proportion histogram, find top marker genes for each cell type with feature plot and dotplot, feature plots for marker genes based on prior knowledge (skin_cluster_marker_panel.png), and motif analysis. After running, 'final.RDS' will be saved for following analysis.

```{r,echo=F,message=F,warning=F}
library(yaml)
library(config)

dir <- params$dir_config
con <- file(dir)
config <- yaml.load(readLines(con))

dir_celltype = config$default$dir_celltype
adjust.cov = config$default$adjust.cov
test_option1 = config$default$test_option1
cell_type_order = config$default$cell_type_order
condition_order = config$default$condition_order
output_dir = config$default$output_dir
dir_after_link_peaks = paste(output_dir,'/','after_link_peaks.RDS', sep="")

library(Seurat)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(dplyr)
library(ggplot2)
library(SoupX)
library(Matrix)
library(scDblFinder)
library(BSgenome.Hsapiens.UCSC.hg38)
library(chromVAR)
library(JASPAR2020)
library(TFBSTools)
library(motifmatchr)
library(stringr)
library(gridExtra)
library(grid)
library(knitr)
library(data.table)
```


## Check before analysis 

Firstly we need to check whether the celltype annotations are generated based on clusters_markers from previous processes or not.

```{r,echo=F,message=F,warning=F}
gene.multiome.seurat.filtered <- readRDS(dir_after_link_peaks)
adjust.cov <- unlist(strsplit(adjust.cov, ","))

a <- as.matrix(read.table(dir_celltype, sep = ",", header = T))

(rowname.check <- all(colnames(gene.multiome.seurat.filtered) == a[, 1]))

if (rowname.check == FALSE) {
  print("Warning: the barcodes are not consistant.")
}

gene.multiome.seurat.filtered$cell.type <- a[, 2]

cell_type_order = unlist(strsplit(cell_type_order, ","))
condition_order = unlist(strsplit(condition_order, ","))
```


```{r,echo=F,message=F,warning=F}
gene.multiome.seurat.filtered$ctc <- paste(gene.multiome.seurat.filtered$Condition,gene.multiome.seurat.filtered$cell.type,sep="__")
```

## UMAP

```{r,echo=F,message=F,warning=F,fig.height=4, fig.width=12.5}
if (length(cell_type_order)!=0) {
  gene.multiome.seurat.filtered$cell.type <- factor(x = gene.multiome.seurat.filtered$cell.type, levels = cell_type_order)
}

grid_arrange_shared_legend <- function(..., nrow = 1, ncol = length(list(...)), position = c("bottom", "right")) {

  plots <- list(...)
  position <- match.arg(position)
  g <- ggplotGrob(plots[[1]] + theme(legend.position = position))$grobs
  legend <- g[[which(sapply(g, function(x) x$name) == "guide-box")]]
  lheight <- sum(legend$height)
  lwidth <- sum(legend$width)
  gl <- lapply(plots, function(x) x + theme(legend.position = "none"))
  gl <- c(gl, nrow = nrow, ncol = ncol)

  combined <- switch(position,
                     "bottom" = arrangeGrob(do.call(arrangeGrob, gl),
                                            legend,
                                            ncol = 1,
                                            heights = unit.c(unit(1, "npc") - lheight, lheight)),
                     "right" = arrangeGrob(do.call(arrangeGrob, gl),
                                           legend,
                                           ncol = 2,
                                           widths = unit.c(unit(1, "npc") - lwidth, lwidth)))
  grid.newpage()
  grid.draw(combined)

}


p1 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.pca", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'cell.type') + ggtitle(paste("RNA (harmony) by cell"))
p2 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.lsi", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'cell.type') + ggtitle(paste("ATAC (harmony) by cell"))
p3 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.wnn", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'cell.type') + ggtitle(paste("WNN (harmony) by cell"))

grid_arrange_shared_legend(p1, p2, p3, nrow = 1, ncol = 3,position="right")

p1 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.pca", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'Condition') + ggtitle(paste("RNA (harmony) by condition"))
p2 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.lsi", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'Condition') + ggtitle(paste("ATAC (harmony) by condition"))
p3 <- DimPlot(gene.multiome.seurat.filtered, reduction = "umap.harmony.wnn", label = TRUE, label.size = 2.5, repel = TRUE, group.by= 'Condition') + ggtitle(paste("WNN (harmony) by condition"))

grid_arrange_shared_legend(p1, p2, p3, nrow = 1, ncol = 3,position="right")
```

## Cell type proportion histogram

```{r,echo=F,message=F,warning=F,fig.height=6, fig.width=10}
#--a:celltype annotation 
tab <- as.data.frame(a)
tab$condition <- gene.multiome.seurat.filtered$Condition
colnames(tab) <- c("barcodes", "cell_type", "condition")

if (length(condition_order)!=0) {
  tab$condition <- factor(x = tab$condition, levels = condition_order)
}
#cell type proportion histogram

if (length(cell_type_order)!=0) {
  tab$cell_type <- factor(x = tab$cell_type, levels = cell_type_order)
}
celltype_sum = tab %>% group_by(condition, cell_type) %>% summarise(n = n())
condition_sum = tab %>% group_by(condition) %>% summarise(n = n())
final = merge(condition_sum, celltype_sum, by = "condition", all.x = T)
final$prop = final$n.y / final$n.x
print(ggplot(final, aes(fill = cell_type, y = prop, x = condition)) + geom_bar(position = "stack", stat = "identity"))
```



```{r,echo=F,message=F,warning=F,fig.height=6, fig.width=10}
#for each sample
tab <- as.data.frame(a)
tab$condition <- gene.multiome.seurat.filtered$Condition
colnames(tab) <- c("barcodes", "cell_type", "condition")
sample <- strsplit(as.character(tab[, 1]), "-")
sample <- as.data.frame(matrix(unlist(sample), ncol = 2, byrow = T))
sample <- sample[, 2]
tab$sample <- sample
tab$sample <- sprintf("sample%s", tab$sample)

if (length(cell_type_order)!=0) {
  tab$cell_type <- factor(x = tab$cell_type, levels = cell_type_order)
}

sample_order = c()

for (i in (1:length(unique(gene.multiome.seurat.filtered$sample)))){
  sample_order=c(sample_order,paste("sample",i,sep=""))
}

tab$sample <- factor(x = tab$sample, levels = sample_order)

celltype_sum = tab %>% group_by(sample, cell_type) %>% summarise(n = n())
sample_sum = tab %>% group_by(sample) %>% summarise(n = n())
final = merge(sample_sum, celltype_sum, by = "sample", all.x = T)
final$prop = final$n.y / final$n.x
print(ggplot(final, aes(fill=cell_type, y=prop, x=sample)) + geom_bar(position="stack", stat="identity") + theme(axis.text.x = element_text(angle = 45, hjust=1)))
```


```{r,echo=F,message=F,warning=F,fig.height=6, fig.width=6}
tab <- as.data.frame(a)
tab$condition <- gene.multiome.seurat.filtered$Condition
colnames(tab) <- c("barcodes", "cell_type", "condition")
sample <- strsplit(as.character(tab[, 1]), "-")
sample <- as.data.frame(matrix(unlist(sample), ncol = 2, byrow = T))
sample <- sample[, 2]
tab$sample <- sample
tab$sample <- sprintf("sample%s", tab$sample)

if (length(cell_type_order)!=0){
tab$cell_type <- factor(x = tab$cell_type, levels = cell_type_order)
}

#split by condition

con<-unique(tab$condition)

for (i in (1:length(unique(tab$condition)))){
  
  cat(sprintf("\nThe following histogram is for %s",con[i]))
  
  temp.sam <- tab %>% filter(condition == con[i])
  
  celltype_sum = temp.sam %>% group_by(sample, cell_type) %>% summarise(n = n())
  sample_sum = temp.sam %>% group_by(sample) %>% summarise(n = n())
  final = merge(sample_sum, celltype_sum, by = "sample", all.x = T)
  final$prop = final$n.y / final$n.x
  print(ggplot(final, aes(fill=cell_type, y=prop, x=sample)) + geom_bar(position="stack", stat="identity")+ggtitle(paste("Histogram for ",con[i]))+ theme(axis.text.x = element_text(angle = 45, hjust=1)))
  
}
```


## marker genes for annotated cell types (Type ONE)

The most upregulated genes from each cluster are the one with the lowest p_val_adj then highest avg_log2FC.

```{r,echo=F,message=F,warning=F,fig.height=26, fig.width=30}
## marker genes for annotated cell types using FindAllMarkers
if (length(cell_type_order)!=0) {
  gene.multiome.seurat.filtered$cell.type <- factor(x = gene.multiome.seurat.filtered$cell.type, levels = cell_type_order)
}

Idents(gene.multiome.seurat.filtered) <- gene.multiome.seurat.filtered$cell.type

DefaultAssay(gene.multiome.seurat.filtered) <- "SCT"


SSc.gene.clustermarkers2 <-
    FindAllMarkers(
      gene.multiome.seurat.filtered,
      only.pos = TRUE,
      min.pct = 0.1,
      logfc.threshold = 0.25,
      test.use = test_option1,
      latent.vars = adjust.cov
    )


#print out the whole test result
write.csv(
  SSc.gene.clustermarkers2,
  paste(output_dir,'/','gene markers for each cell type_RNA.csv',sep=""),
  row.names = F,
  quote = F
)

#Choose the topmarker for each cell type:

markers <-
  SSc.gene.clustermarkers2 %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC), .by_group = TRUE) %>% dplyr::slice(1)

write.csv(markers, file = paste(output_dir,'/',"rna_top_gene_celltype.csv",sep=""))

markers <-
  SSc.gene.clustermarkers2 %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC)) %>% dplyr::slice(1) %>% pull(gene)

print("The following shows the topmarker gene for each cell type:")


name <- unique(SSc.gene.clustermarkers2$cluster)
tab.1 <- cbind(as.vector(name), markers)
colnames(tab.1) <- c("cell type", "markers")
tab.1 <- setDT(as.data.frame(tab.1))
kable(tab.1)

print("The following is the Vlnplots of topmarkers:")

VlnPlot(gene.multiome.seurat.filtered, features=markers, ncol=4)

```


The following is the feature plot of the five most upregulated gene from each cluster with the lowest p_val_adj then highest avg_log2FC:


```{r,echo=F,message=F,warning=F}
#choose the top 5 genes for each cell type
markers <- SSc.gene.clustermarkers2 %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC), .by_group = TRUE) %>% dplyr::slice(1:5)

cell.name <- as.vector(markers %>% pull(cluster) %>% unique())

gene<-markers[,7]

for (i in (1:length(levels(gene.multiome.seurat.filtered)))){
  
  print(paste("Feature plot for ",cell.name[i],":",sep=""))
  
  print(FeaturePlot(gene.multiome.seurat.filtered, features = gene[[1]][(1:5)+(i-1)*5],reduction="umap.harmony.wnn"))
}
```


```{r,echo=F,message=F,warning=F,fig.height=16,fig.width=12}
## Dotplot of topmarkers using FindAllMarkers

temp.genes <-
  SSc.gene.clustermarkers2 %>% group_by(cluster) %>% arrange(p_val_adj, desc(avg_log2FC), .by_group = TRUE) %>% dplyr::slice(1:5) %>% pull(gene) %>% unique()

print("The following is the Dotplots of topmarkers:")

tempplot <- DotPlot(gene.multiome.seurat.filtered, assay= "SCT", features=temp.genes, cols=c("lightgrey","navy"),group.by= 'cell.type')

tempplot + theme(axis.text.x = element_text(angle = 45, hjust=1))+coord_flip()

saveRDS(gene.multiome.seurat.filtered,file = paste(output_dir,'/','before_motif.RDS', sep=""))
```



## Motif 

```{r,echo=F,message=F,warning=F}
# Scan the DNA sequence of each peak for the presence of each motif, and create a Motif object
# https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis.html

pwm_set <- getMatrixSet(x = JASPAR2020, opts = list(species = 9606, all_versions = FALSE))

### gene
DefaultAssay(gene.multiome.seurat.filtered) <- "peaks"

### motif x feature

temp.motif.matrix <- CreateMotifMatrix(features = granges(gene.multiome.seurat.filtered), pwm = pwm_set, genome = 'BSgenome.Hsapiens.UCSC.hg38', use.counts = FALSE)

temp.motif.object <- CreateMotifObject(data = temp.motif.matrix, pwm = pwm_set)

gene.multiome.seurat.filtered <- SetAssayData(gene.multiome.seurat.filtered, assay = 'peaks', slot = 'motifs', new.data = temp.motif.object)

### get chromatin accessibility motif activities

gene.multiome.seurat.filtered <- RunChromVAR(object = gene.multiome.seurat.filtered,  genome = BSgenome.Hsapiens.UCSC.hg38)

### add motif sequence information ### might not need this, as the above script has already added motif !!!!!!!!!!! see pbmcs

gene.multiome.seurat.filtered <- AddMotifs(gene.multiome.seurat.filtered, genome= BSgenome.Hsapiens.UCSC.hg38, pfm=pwm_set)
```

## identify marker peaks based on each cell type (Type One)

```{r,echo=F,message=F,warning=F}
### identify marker peaks

### gene
DefaultAssay(gene.multiome.seurat.filtered) <- "peaks"
Idents(gene.multiome.seurat.filtered) <- gene.multiome.seurat.filtered$cell.type


DE.Args <-
  rownames(table(gene.multiome.seurat.filtered$cell.type))
cell.matrix.atac <- c()

for (i in DE.Args) {
   
   adjust <- c('nCount_peaks',adjust.cov)
   temp.ac <-
      FindMarkers(
        gene.multiome.seurat.filtered,
        ident.1 = i,
        min.pct = 0.05,
        test.use = test_option1 ,
        logfc.threshold = 0.25,
        latent.vars = adjust
      )
  
  if (nrow(temp.ac)==0){
    print(paste("for",i,", no features pass logfc.threshold threshold"))
  }
  else{
    temp.ac <-
      cbind(temp.ac, i)
    colnames(temp.ac)[6] <- "cluster"
    temp.ac <- cbind(rownames(temp.ac), temp.ac)
    colnames(temp.ac)[1] <- "Chromatin"
    temp <- temp.ac
    cell.matrix.atac <- rbind(cell.matrix.atac, temp)
  }

}


# Finally we collapse the list of hits down to a single table and sort it by FDR to put the most significant ones first

write.csv(
  cell.matrix.atac,
  paste(output_dir,'/',"gene markers for each cell type_ATAC.csv", sep=""),
  row.names = F,
  quote = F
)

SSc.gene.clusterpeaks<-cell.matrix.atac %>% arrange(p_val_adj)

SSc.gene.clusterpeaks$p_val <- as.numeric(SSc.gene.clusterpeaks$p_val)

# get sig. differentially accessible peaks; FDR <= 0.1 does not have large power
SSc.gene.clusterpeaks  %>% filter(p_val_adj<=0.01 & avg_log2FC>=0.25) -> SSc.gene.clusterpeaks.sig
```


```{r,echo=F,message=F,warning=F}
### motif enrichment analysis

### gene

### retrieve a set of background peaks
### Matching the set of background peaks is essential when finding enriched DNA sequence motifs. By default, we choose a set of peaks matched for GC content, but it can be sometimes be beneficial to further restrict the background peaks to those that are accessible in the groups of cells compared when finding differentially accessible peaks.
### The AccessiblePeaks() function can be used to find a set of peaks that are open in a subset of cells. We can use this function to first restrict the set of possible background peaks to those peaks that were open in the set of cells compared in FindMarkers(), and then create a GC-content-matched set of peaks from this larger set using MatchRegionStats().
# find peaks open in Pvalb or Sst cells
temp.bg.peaks <- AccessiblePeaks(gene.multiome.seurat.filtered)

# match the overall GC content in the peak set
temp.meta.feature <- GetAssayData(gene.multiome.seurat.filtered, assay = "peaks", slot = "meta.features")

### loop through each cell type for the cluster peaks to do matching and motif enrichment

enriched.gene.motifs <- list()

### use significant peaks, not all peaks!!!
for (c in unique(SSc.gene.clusterpeaks.sig$cluster)){
  print(c)
  tempquery <- (SSc.gene.clusterpeaks.sig %>% filter(cluster==c))$Chromatin
  temp.peaks.matched <- MatchRegionStats(  meta.feature =temp.meta.feature[temp.bg.peaks, ],  query.feature = temp.meta.feature[tempquery, ],  n = 50000)
  
  ## Matching GC.percent distribution, peaks.matched can then be used as the background peak set by setting background=peaks.matched in FindMotifs().
  # test enrichment
  enriched.gene.motifs[[c]] <- FindMotifs(object = gene.multiome.seurat.filtered,  features = tempquery, background=temp.peaks.matched)
  
}

### put that as tibble
sapply(names(enriched.gene.motifs),function(x) {
  enriched.gene.motifs[[x]]$cluster <<- x
})
# Finally we collapse the list of hits down to a single table and sort it by FDR to put the most significant ones first
as_tibble(do.call(rbind, enriched.gene.motifs)) %>% arrange(pvalue) -> enriched.gene.motifs

enriched.gene.motifs$padj <- p.adjust(enriched.gene.motifs$pvalue,method="fdr")
```

The following is motif plots the top 5 TFs:

```{r,echo=F,message=F,warning=F,fig.height=4,fig.width=8}
### motif plots the top 5 TFs

for (c in unique(enriched.gene.motifs$cluster)){
  
  tempmotifs <- enriched.gene.motifs %>% filter(cluster==c) %>% dplyr::slice(1:5)  %>% pull(motif)
  
  print(MotifPlot(gene.multiome.seurat.filtered,motifs=tempmotifs, assay="peaks")+ggtitle(c))
  
}

saveRDS(gene.multiome.seurat.filtered,file=paste(output_dir,'/',"final.RDS", sep=""))
```


```{r,echo=F,message=F,warning=F}
# The following is used to check if sig. motifs are also sig. expressed in the corresponding cell type

#visualize motif activities per cell, and also provides an alternative method of identifying differentially-active motifs between cell types. ChromVAR identifies motifs associated with variability in chromatin accessibility between cells. We can also directly test for differential activity scores between cell types. This tends to give similar results as performing an enrichment test on differentially accessible peaks between the cell types (shown above). 

### check if sig. motifs are also sig. expressed in the corresponding cell type
enriched.gene.motifs %>% filter(fold.enrichment>=2, padj<=0.01) -> enriched.gene.motifs.sig

enriched.gene.motifs.sig$MotifName <- ConvertMotifID(gene.multiome.seurat.filtered, id=enriched.gene.motifs.sig$motif)

enriched.gene.motifs.sig$MotifName <- gsub("\\(var.3)","",gsub("\\(var.2)","",enriched.gene.motifs.sig$MotifName))

saveRDS(enriched.gene.motifs,file=paste(output_dir,'/',"enriched_gene_motifs.RDS", sep=""))
```








